src/api/app/components/watchlist_component.rb

```diff
@@ -11,16 +11,33 @@ class WatchlistComponent < ApplicationComponent
     'BsRequest' => 'Watch this request'
   }.freeze
 
-  def initialize(user:, project: nil, package: nil, bs_request: nil)
+  def initialize(user:, bs_request: nil, package: nil, project: nil)
     super
 
     @user = user
-    # NOTE: the order of the array is important, when project and package are both present we ensure it takes package.
-    @object_to_be_watched = [bs_request, package, project].compact.first
+    @object_to_be_watched = object_to_be_watched(bs_request, package, project)
   end
 
   private
 
+  def object_to_be_watched(bs_request, package, project)
+    return bs_request if bs_request
+
+    # this is a remote project, don't offer watching.
+    return unless project.is_a?(Project)
+
+    # there is no package, offer watching the project
+    return project unless package
+
+    # maybe package is a multibuild flavor? Try do look up the object of the flavor.
+    package = Package.get_by_project_and_name(project, package, { follow_multibuild: true }) if package.is_a?(String)
+
+    # the package is coming via a project link, don't offer watching it.
+    return if package.project != project
```

Comment:
Yes. My suggestion prevents from throwing the exception in the view component. In case a package is not found, don't throw an exception: just don't offer to add/remove a watchlist item.

Reply:
> My suggestion prevents from throwing the exception

not really, your suggestion is way past line 33 right? :) The only way to prevent this exception thrown by `Package.get_by_project_and_name` (which will just lead to a 404) would be to catch it...

Reply:
When something doesn't exist we should not return an exception...

Reply:
we don't, the exception turns into a 404.

Reply:
Ah, yes. No need to change the code then.